### DISCUSSION TITLE
Rolling static keys/rekeying for anonymity.

### DISCUSSION THREAD

**Discussion by ellenhp** [markqvist/Reticulum#376] [General]
Hey, I'm wondering if there's been any thought put into the threat model of someone (let's say a state actor) monitoring an area for announce packets to keep track of who's present in an attempt to crack down on freedom of assembly. Not that any state actor has ever done that, but it might be something worth considering anyway.

The easiest way to handle this would be at the application level, by rotating through different identities/destinations and burning old ones, but without some cooperation at the protocol level you'd have to choose between blowing your cover by announcing an old destination with 0 hops, or risking non-delivery of packets addressed to your previous destinations.

I could imagine something as sophisticated as  constructing an arbitrary social graph for your spoofed identities and setting the number of announced hops appropriately for the one you're announcing as. For added confusion, you could even exchange unused spoofed identities with trusted parties to subvert any attacker's attempts at clustering of identities by announced hop count, as the social graph would change dramatically if you exchanged an identity with a friend. As far as I can tell there are no facilities for this, but it's something I'd like to see. I don't think that a LoRa PHY layer can tolerate much chaff with its bandwidth constraints, but Reticulum or LXMF should probably have a story here because this is a legitimate threat in way too many places.

**Comment by faragher** [comment]
The ability to switch or discard identities is core to Reticulum. Rolling through them is generally as simple as switching out a file. Application-specific rolling identities is possible, but as you say, is application specific. There's no practical reason you can't make an LXMF client that aggregates multiple identities into a single message stream and sends to a random member on a list.

However, that's just "security through obscurity." While deception is a part of a layered defense, it's really just playing for time. A holistic approach to network security (in both pertinent definitions of "network") is warranted and, quite frankly, I should sit down and consider the situation more thoroughly. I don't have this library for show. :D

But it's worth a simple example to assess the threat posed by such an attack. Yes, this will get very long. I'm sorry.

Alice and Bob do not wish to be detected. Eve is a state actor hostile to Alice and Bob. There are three public distribution nodes, Alpha, Bravo, and Charlie, all secure, as well as Eve's node, Echo. They all see each other and have arbitrary interfaces attached. 

Echo (and thus Eve) can see Alice and Bob's announces, and see they're two hops away, as well as if they're through A, B, or C, and thus Eve knows they're connected to those nodes. If their identities are compromised, that's a problem, but if it's a fresh identity and a widely used node, then there should be no need for suspicion. Alice sends a message to Bob. Eve never sees it, since it's not routed through Echo. Bob to Alice is the same.

Eve compromises Alpha. which is where Alice and Bob are connected. Alice sends Bob a message. Eve sees that Bob is receiving a message, but not that Alice sent it. Time of fight analysis can be an issue, but given an active node with multiple actors, it's difficult to determine who's talking to whom (assuming automatic receipts are off). 

Bob moves to Bravo. Now Eve sees Bob's packets go to Bravo, but without an announce has no idea where they go from there. 

Alice starts node Delta, attached only to Charlie, and now Eve sees announces come from Charlie, but three hops away. This means any node connected to Charlie could be the next step in the chain. Eve only knows that Alice's packets are going to Bob and routed through Delta, not who sent them or their exact node of origin.

Alice uses two identities, one attached to Alpha and the other to Delta. Both send to Bob, and it's impossible to link the two identities to the same sender without out of band knowledge.

A fully crosslinked network, while important for distribution at our current scale, is less secure than people spinning up their own node and linking to a single endpoint. This creates a fault intolerant network, but it also creates a path with multiple hops and added security. The more people run their own nodes, the more secure the network becomes. (at least in this threat case) If something is three hops away, and the node two hops away has five nodes attached, an announce leaks very little information.

It's also helpful to run a node with IFAC enabled, effectively making a password protected interface, so that node will always be one hop away from any malicious actor.

It might be worth the ability to hard-code or ingest paths and public keys using out of band communications, such as paper messages, but I think the threat of this kind of announce-based tracking is a wider operational security issue, where a healthy ecosystem provides all the protection necessary so long as the identity doesn't somehow out itself as suspicious. 

I believe it to be an issue of network topology and how people intend to use the network as opposed to an engineering issue. If nations A and B don't want communication, then both nodes connecting to nation C completely obfuscates crossing the border. A only knows B is two hops away, not where the server is.

If there's a specific case I'm overlooking, please let me know, but I think auto-rotating identities due to announces isn't the best option.

**Reply by ellenhp** [reply]
The use case I have in mind is a mobile device that may communicate over some sort of short-range wireless link, and the threat is that Eve is colocated with both Alice and Bob. Alice and Bob may even be able to directly communicate, but as soon as an announce is emitted from one of their radios, letting the other know that they're around, Eve knows that whoever announced was present on that particular day. If that happens two days with the same static key, anonymity becomes pseudonymity and the risks to their safety begin to grow.

IMSI catchers are normally what are used to execute this attack, but it's much much easier to do with reticulum than cellular. Or rather the hardware is generally cheaper and more accessible. Hopefully that makes sense. I understand that reticulum over other transports is probably more resilient to this attack, but this is definitely something that should be figured out, probably at the LXMF level.

**Reply by faragher** [reply]
If the network is closed, you can protect the network with an IFAC and the announces won't leak (outside of radio direction-finding, which will reveal the sender's location, but not an identity. 

If you have a closed network and an uplink to an open network, you can uplink to another trusted server, and then you'd be a minimum of two hops away on a network that has an unknown number of connections from unknown locations. Over LoRa or similar radio, they may be able to sniff the radio itself. but you can announce outside of the area under surveillance and have an announce rate lower than the time you expect to be in the area. 

You can also change between full, AP, roaming, or boundary mode to keep announces within a subnet or otherwise work around keeping your announces away from unauthorized users. 

For maximum security you can set a new identity per day, announce before traveling to a hostile area, and burn the identity at end of day. You can manually change your identity by going to the Encryption Keys page. This is functionally identical to changing the IMEI, Bluetooth address, and subscriber number on a phone, changing all identifying information as well as having new encryption keys. It's actually better, since there's no other metadata leakage. While I understand this is a manual process and not a rotating or multi-channel system, it does provide security with a currently available app.

Remember: If the announce reaches the propagation node, the other radios don't need to see it. They can simply query the network for a path, meaning they can announce asynchronously outside of an area where eavesdroppers are an issue.

Be aware a Bluetooth radio being on is more dangerous, and you'll need to use wired systems.

**Reply by ellenhp** [reply]
> Remember: If the announce reaches the propagation node, the other radios don't need to see it. They can simply query the network for a path, meaning they can announce asynchronously outside of an area where eavesdroppers are an issue.

Oooh, that's really clever actually. If this is something that needs to be handled at the application level then I'll have to keep that in mind while I'm scoping work. This will be new hardware probably running on an RP2040 or STM32F4 or something like that. Level of effort required for an embedded port of reticulum seems... high. I [started it](https://github.com/ellenhp/reticulum-rs) but I'm not sure I'll be able to get it to a reasonable state before I need to move on to other aspects of this project. I want to start work on the schematic this weekend :weary: 

I'm also looking into building something on top of `noise` since there's already a rust crate out there for it that can be easily ported to run on microcontrollers, and I think it might suit my needs pretty well. It would be missing a lot of reticulum's more advanced features, but it might be easier and less stressful to use that than rewriting all of those advanced features for my own use. Using cryptographic primitives at this low a level feels like playing with fire.

Thank you for the input! This makes sense to me but leaves me with a lot to think about.

**Reply by cobraPA** [reply]
Well, this seems kind of an odd way to think about working in a hostile environment.  If you are trying to survive in a hostile environment then wouldn't you turn off announces?  Keys should be exchanged out-of-band, otherwise you have no guarantee of who you are talking to.  Once you have the key to guarantee your target for communication, then you will be anonymous.

**Reply by ellenhp** [reply]
I don't think it's that weird at all. Am I wrong in my assumption that a client would not transmit a message indended for me unless I announce my presence?

My assumptions:

* There is no relay node that both Alice and Bob both consistently have access to
* There is no location that both Alice and Bob regularly meet that isn't surveilled
* There is no path through the network other than the direct path e.g. because of an internet shutdown.

If clients are really just flooding the network with messages regardless of whether their destinations have announced themselves then I don't think this protocol is very bandwidth efficient at all.

Edit: And if they aren't flooding (i.e. if my assumption earlier is correct) then turning off announces is equivalent to not communicating. Which is not the worst failure mode, but I would hope for better.

**Reply by ellenhp** [reply]
Sorry, I have some more thoughts.

> Keys should be exchanged out-of-band, otherwise you have no guarantee of who you are talking to.

Not really feasible much of the time because of the length of a key. However, as long as you *verify* the keys OOB, or do something akin to a Signal safety number, you should be fine.

> Once you have the key to guarantee your target for communication, then you will be anonymous.

This is not true, and it is IMO very important to be precise with language when you're dealing with technology like this. Reticulum packets (header1/header2, not sure about the lrproof from memory) have a destination address/truncated hash in their header in the clear. If you are Eve and you *ever* see a destination address that you know from a previous encounter, that destination is no longer anonymous, they are pseudonymous. I know this seems pedantic but it is absolutely vital for a lot of practical de-anonymization attacks to make this distinction, because once you have a pseudonym, you can begin correlating other data, like license plate numbers, usages of a particular bus fare card, things like that. There's no silver bullet here because destination address reuse is very difficult to avoid, even potentially impossible practically speaking, but reusing the same destination address for too long is dangerous.

**Reply by faragher** [reply]
Basically in reverse:

>If clients are really just flooding the network with messages regardless of whether their destinations have announced themselves then I don't think this protocol is very bandwidth efficient at all.

You are correct that's extremely inefficient, and it is not the way Reticulum acts (see below)

> My assumptions:
> 
> * There is no relay node that both Alice and Bob both consistently have access to
> * There is no location that both Alice and Bob regularly meet that isn't surveilled
> * There is no path through the network other than the direct path e.g. because of an internet shutdown.

Noted. Point one needs some further discussion, but we will work with the other two.

>I don't think it's that weird at all. Am I wrong in my assumption that a client would not transmit a message indended for me unless I announce my presence?

You are partially correct. A message requires two things, a path and encryption keys, both of which are contained in an announce. If a destination never announces, given your above criteria, it will never receive a message.

However, that doesn't mean you can't send them a message. You can query the network if you don't have that information. This is not automatic, but almost all programs I've seen use it as a fallback behavior. This means the distribution node needs to see the announce, not every client. 

Second, if you can't currently see the other node (due to it being offline or a link in the chain being broken) a trusted distribution node can hold the message. Please note that you need to trust the node manually (or enter it in, say Sideband) and can mark any node or peer untrusted. When the distribution node sees the other party, it will transfer their messages.

More than that, a distribution node will (can?) sync message digests with other nodes, even if they're not constantly connected. So bringing a node into contact with another every day or two will transfer pending messages and slow communication can be performed that way, with mobile, transient nodes and/or a dead drop node. Something thrown in a backpack/purse/pocket and simply walked into range over lunch can thus become a "sneakernet" style message transport system.

So if Alice and Bob can see a node, or are running nodes locally, they can announce privately and query the network for keys later. I am certainly seeing the call for manually entered paths and persistent encryption key storage for EMCON environments, but I don't think that's currently available.

Regarding point one, I'm uncertain if there is a "point to point" mode in Reticulum. A node can be run on very light hardware, but it is basically required. 

Also, if you're looking for unrouted point to point communication, it's possible to use the RNode without Reticulum and simply broadcast arbitrary data. It does end up running into the same problem of needing an "announce" or an out of band encryption key exchange, but it doesn't require pathing information. 

It might be worth reconsidering carrying a full distribution node with you (which could be as small as a Pi Zero and an RNode, although you'll have to compile the crypto libraries yourself or run the pure python version, IIRC. Small problems easily solved) so these exchanges and queries look more like normal network transactions between servers and not an individual broadcaster.

I'd also like to point out using radios in a hostile environment is a bad idea unless you look like normal traffic. If you have two people using encrypted radio, that's an easy catch. The key is to have dozens using radios for legitimate (as defined by the opposition) traffic and 
blend in with the crowd. It might be worth considering using ad-hoc WiFi signals to communicate directly. If you turn off your Bluetooth and cellular data, that shouldn't leave too much of a footprint (assuming they don't access your IMEI with the cell service off). At that point it's worth using a new phone with no SIM and using an ad-hoc WiFi system. Yes, an SBC would be more secure, but nobody questions someone typing away on their phone, and a complicated home-brew electronic box sets off alarms.

That being said, we are now fantastically out of scope for Reticulum. Sorry, I go wandering sometimes.

**Reply by faragher** [reply]
> Reticulum packets (header1/header2, not sure about the lrproof from memory) have a destination address/truncated hash in their header in the clear. If you are Eve and you _ever_ see a destination address that you know from a previous encounter, that destination is no longer anonymous, they are pseudonymous. 

Correct and well said. Countering this is terribly difficult, however. As I said, not getting picked up in the first place is the best solution. Changing identities is painless, except for the announce issue, and is quite functional. Not being seen at all is best, and can be accomplished with secure nodes, but you still leak data over the radio. A packet requires a destination that can be read, and you have to assume the attacker has access to your protocol and source information.

There is, practically, no way to prevent leaking source data. I do, however, have a rather . . . hacky solution for you. This may, in fact, be a very bad idea, but I'll let you decide.

Destination addresses are predictable in one direction. I'd need to check if announces are by identity or by address, but you could change aspects to completely alter your address. Obviously you'd need some kind of pre-arranged system, but if you keep your base identity hidden, then, say, LXMF.Message.JAN1 and LXMF,Message.JAN2 are both deterministic using a one way function, but cannot be traced back to the original identity (to a cryptographic definition of "cannot") Both use the same encryption and decryption key, but have a vastly different destination hash. If correlation of destination hashes is your primary concern, this can be an option (although it will require solid planning, a custom program, and some operational support). 

I encourage you to look at your threat matrix again and consider if you're designing around a need or around a specification. No judgment, it's just we're really getting into the weeds.

**Reply by ellenhp** [reply]
My reticulum implementation doesn't need a SBC. It runs on bare metal. :) So yes I suppose I'm talking about running lightweight nodes, in reticulum parlance. But my implementation isn't complete and I'm trying to evaluate whether I should complete it. I'm very indecisive about it. It's a pretty heavyweight protocol and if it doesn't have a first-class way to obfuscate the identifiers I'm worried about then that's a real bummer.

I've witnessed with my own eyeballs situations with threat models like the one I'm describing, so I do think it's worth thinking about and that I'm not prematurely optimizing or anything like that. It's very easy to get caught up in cryptography-land and forget about the real world, but there are lots of situations where Alice and Bob may be able to use radios in a hostile area without fear for their immediate physical safety, but safety the next morning gets very complicated if they leak too much information.

> Destination addresses are predictable in one direction. I'd need to check if announces are by identity or by address, but you could change aspects to completely alter your address. Obviously you'd need some kind of pre-arranged system, but if you keep your base identity hidden, then, say, LXMF.Message.JAN1 and LXMF,Message.JAN2 are both deterministic using a one way function, but cannot be traced back to the original identity (to a cryptographic definition of "cannot") Both use the same encryption and decryption key, but have a vastly different destination hash. If correlation of destination hashes is your primary concern, this can be an option (although it will require solid planning, a custom program, and some operational support).

This is again extremely clever. I'll need to read a bit too before I say something incorrect but I think this might just solve the problem entirely. I can't remember if announces include public keys in the clear though, as that would break this sort of obfuscation I think.

**Reply by faragher** [reply]
Yeah, I've seen too many people hyper-focus on one aspect rather than look at the entirety of the situation. Tracking metadata, targeting based on signatures, and network-centric operations are all pretty advanced in implementation, and I respect that you're addressing the whole problem. That being said, it's far easier to have an engineering solution than expect peak performance in a dynamic and high-fatigue environment. 

The complication here is your requirements. You could have an end to end encrypted network with a single pre-shared key or even a one-time pad if you had the ability to pre-share information, or dynamically announce using Reticulum prior to entering the AO. Using a relatively rare protocol is suspicious on its face, though. It's a complicated problem, as you know.

If you store an identity's key for a long enough period, you could simply announce only once and prevent leaking your identity that way. 

I'm uncertain Reticulum meets your needs, but I'm willing to keep trying if you are. :D

**Reply by ellenhp** [reply]
> I'm uncertain Reticulum meets your needs, but I'm willing to keep trying if you are. :D

Same. I was hoping to avoid reinventing the wheel though! I wrote this spec up for something that might work better for my use-case. Currently trying to nerd snipe people into reviewing it/breaking it but fair warning: It's quite long. :wink:  I think it has much lower overhead than reticulum, just 1 byte of header, an 8 byte message ID and 16 byte MAC per message once a channel is established. Establishing a channel takes <150 bytes exchanged and is 0-RTT. No public keys are transmitted, nor signatures, and there aren't any addresses or announces. It's a weird idea but I think it'll work. None of my friends have been able to break it yet but I'm sure it has flaws. It's not really a replacement for reticulum because it uses spray-and-wait instead of announces, meaning it's impossible to construct a routing table and message delivery rate will suffer because of that. But I just thought I'd close the loop here, because I did end up having to reinvent the wheel.

https://gist.github.com/ellenhp/149a6e3368f54917aab294375b7321dc

**Reply by faragher** [reply]
Based off a cursory look at your spec, it would prevent destination leakage at the cost of large-scale routability and increased raw number of transmissions. From a trilateration standpoint, or building a TEMPEST profile, you'd increase the risk of correlation with video or other detection mechanisms, as well as broadcasting your position much more frequently. As an attacker, I'd pivot to signatures, watching BT addresses and the like when a broadcast was detected and checking both transport and choke points near the times the transmissions started and ended.

However, as the old maxim goes, if you want to know if your door and windows are secure, ask a firefighter how they'd get in. When they tell you using an axe to go through the wall, your work is done.

**Reply by ellenhp** [reply]
Thanks for taking a look! And yeah if we're talking about physical link signatures or multilateration then I think my work at the protocol level is done. Routability outside of a few hops is going to be somewhere between inconsistent and terrible, but I may consider extending the protocol at some point to improve that, or just add optional support for a different protocol to the devices I'm building. I think the bundle protocol v7 could be better for reachability, or maybe I'll finish my reticulum port :)
